### PR TITLE
Bug Bash and Full Code Review of Java Client

### PR DESCRIPTION
**Pull request by mfateev**
First focus on public API which is all packages that are not under com.uber.cadence.internal.
But any comments on implementation are also very welcome.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,534 @@
...
+It can also answer to synchronous queries and receive external events (aka signals).
+
+## Workflow Interface
+
+A workflow must define an interface class. All its methods must have one of the following annotations:
+- @WorkflowMethod indicates an entry point to a workflow. It contains parameters like timeouts and task list. Required
+parameters (like executionStartToCloseTimeoutSeconds) that are not specified through the annotation must be provided at runtime.
+- @Signal indicates a method that reacts to external signals. Must have a `void` return type.
+- @Query indicates a method that reacts to synchronous query requests.
+It is possible to have more than method with the same annotation.
```
Comment (vancexu) [inline README.md]: missing word "more than method"

File: README.md [hunk 2]
```diff
@@ -1 +1,534 @@
...
+### Calling Activities
+
+`Workflow.newActivityStub` returns a client side stub that implements an activity interface. 
+It takes activity type and activity options as arguments. Activity options are needed only if some of the required
+ timeouts are not specified through @ActivityMethod annotation.
+
+Calling a method on this interface invokes an activity that implements this method. 
+An activity invocation synchronously blocks until the activity completes (or fails or times out). Even if activity 
+execution takes a few months the workflow code still see it as a single synchronous invocation.
+Isn't it great? I doesn't matter what happens to the processes that host the worklfow. The business logic code
```
Comment (vancexu) [inline README.md]: typo

File: README.md [hunk 3]
```diff
@@ -1 +1,534 @@
...
+Given a workflow interface executing a workflow requires initializing a `WorkflowClient` instance, creating 
+a client side stub to the workflow and then calling a method annotated with @WorkflowMethod.
+```java
+WorkflowClient workflowClient = WorkflowClient.newClient(cadenceServiceHost, cadenceServicePort, domain);
+// Create workflow stub
+FileProcessingWorkflow workflow = workflowClient.newWorkflowStub(FileProcessingWorkflow.class);
+```
+There are two ways to start workflow execution. Synchronously and asynchronously. Synchronous invocation starts a workflow
+and then waits for its completion. If process that started workflow crashes or stops waiting workflow continues execution.
+As workflows are potentially long running and crashes of clients happen it is not very commonly found in production use.
```
Comment (vancexu) [inline README.md]: Please add punctuation to make it easier to understand.

File: README.md [hunk 4]
```diff
@@ -1 +1,534 @@
...
+        return greeting.get();
+    }
+}
+```
+Calling methods annotated with @QueryMethod is not allowed from within a workflow code.
+### Workflow Implementation Constraints
+
+Cadence uses [event sourcing](https://docs.microsoft.com/en-us/azure/architecture/patterns/event-sourcing) to recover 
+the state of a workflow object including its threads and local variable values.
+. In essence every time a workflow state has to be restored its code is reexecuted from the beginning ignoring side
```
Comment (vancexu) [inline README.md]: additional "."

Comment (vancexu) [reply README.md]: please also add punctuation to make this long sentence easier to read

File: src/main/java/com/uber/cadence/worker/Worker.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+     * <p>
+     * Implementations that share a worker must implement different interfaces as a workflow type
+     * is identified by the workflow interface, not by the implementation.
+     * </p><p>
+     * The reason for registration accepting workflow class, but not the workflow instance is
+     * that workflows are stateful and a new instance is created for each workflow execution.
+     * </p>
+     */
+    public void registerWorkflowImplementationTypes(Class<?>... workflowImplementationClasses) {
+        workflowWorker.setWorkflowImplementationTypes(workflowImplementationClasses);
```
Comment (yiminc) [inline src/main/java/com/uber/cadence/worker/Worker.java]: we check activityWorker == null in registerActivitiesImplementations, should we do that for workflowWorker here?

Comment (yiminc) [reply src/main/java/com/uber/cadence/worker/Worker.java]: and checkStarted() as well?

File: src/main/java/com/uber/cadence/worker/Worker.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+     * Shutdown a worker, waiting for activities to complete execution up to the specified timeout.
+     */
+    public void shutdown(Duration timeout) {
+        try {
+            long time = System.currentTimeMillis();
+            if (activityWorker != null) {
+                activityWorker.shutdownAndAwaitTermination(timeout.toMillis(), TimeUnit.MILLISECONDS);
+            }
+            if (workflowWorker != null) {
+                long left = timeout.toMillis() - (System.currentTimeMillis() - time);
```
Comment (yiminc) [inline src/main/java/com/uber/cadence/worker/Worker.java]: left could be negative if activityWorker took too long to stop.

File: src/main/java/com/uber/cadence/activity/DoNotCompleteOnReturn.java [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+import java.lang.annotation.Target;
+
+/**
+ * Used to annotate activity <b>implementation</b> method to indicate that Cadence activity is not completed
+ * upon method returning. Use {@link com.uber.cadence.client.ActivityCompletionClient} to complete, fail or heartbeat
+ * such activity asynchronously (even from a different process).
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.METHOD)
+public @interface DoNotCompleteOnReturn {
```
[the remark this answers is not in the document]
Comment (mfateev) [reply src/main/java/com/uber/cadence/activity/DoNotCompleteOnReturn.java]: Good point. Converted from annotation to Activity.doNotCompleteOnReturn() API.

File: src/main/java/com/uber/cadence/workflow/Workflow.java [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+     * } catch (Exception e) {
+     *     throw CheckedExceptionWrapper.throwWrapped(e);
+     * }
+     * </pre>
+     * If throwWrapped returned void it wouldn't be possible to write <code>throw CheckedExceptionWrapper.throwWrapped</code>
+     * and compiler would complain about missing return.
+     *
+     * @return never returns as always throws.
+     */
+    public static RuntimeException throwWrapped(Throwable e) {
```
[the remark this answers is not in the document]
Comment (mfateev) [reply src/main/java/com/uber/cadence/workflow/Workflow.java]: Changed to wrap method which returns wrapped exception without throwing.
Exposed unwrap method.

File: src/main/java/com/uber/cadence/worker/Worker.java (hunk 5 above)
Comment (mfateev) [reply src/main/java/com/uber/cadence/worker/Worker.java]: good catch

File: src/main/java/com/uber/cadence/worker/Worker.java (hunk 6 above)
Comment (mfateev) [reply src/main/java/com/uber/cadence/worker/Worker.java]: shutdownAndAwaitTermination handles negative value gracefully.

**Comment by mfateev** [comment]
I'm going to close this pull request as it was created purely for the complete code review.
The following changes addressed the raised issues:

- https://github.com/uber-java/cadence-client/pull/111
- https://github.com/uber-java/cadence-client/pull/112
- https://github.com/uber-java/cadence-client/pull/113
- https://github.com/uber-java/cadence-client/pull/114

**Pull request closed without merging**